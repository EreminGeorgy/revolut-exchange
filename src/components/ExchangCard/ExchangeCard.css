.card-content {
  display: flex;
}

.currency-input.input > input {
  text-align: right;
  width: 12em;
}
