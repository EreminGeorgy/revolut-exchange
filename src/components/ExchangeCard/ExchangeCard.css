.card-controls {
  display: flex;
}

.currency-input.input > input {
  text-align: right;
  width: 12em;
}
