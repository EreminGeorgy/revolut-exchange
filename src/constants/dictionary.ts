export const DICTIONARY = {
  buy: 'Buy',
  sell: 'Sell',
  EUR: 'EUR',
  USD: 'USD',
  GBP: 'GBP'
}