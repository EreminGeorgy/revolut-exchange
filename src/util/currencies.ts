export const formatCurrencyString = (value: string, locale: string): string => {

  const formatter = new Intl.NumberFormat(locale, {
    style: 'currency',
    currency: 'USD',
    currencyDisplay: "code"
  });

  return formatter.format(Number(value)).replace("USD", "").trim()
}