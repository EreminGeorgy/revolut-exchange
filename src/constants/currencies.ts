import { DICTIONARY } from "./dictionary"

const EUR = 'EUR'
const GBP = 'GBP'
const USD = 'USD'

export const CURRENCIES = {
  EUR,
  GBP,
  USD
}

export const currenciesData = [
  { key: EUR, value: EUR, flag: 'eu', text: DICTIONARY.EUR },
  { key: GBP, value: GBP, flag: 'uk', text: DICTIONARY.GBP },
  { key: USD, value: USD, flag: 'us', text: DICTIONARY.USD },
]
