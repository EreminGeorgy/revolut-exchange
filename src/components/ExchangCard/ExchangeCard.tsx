import React from 'react'
import { Card, Dropdown } from 'semantic-ui-react'
import { currencies } from '../../constants/currencies'

const ExchangeCard: React.FC = () => {
  return (
    <Card>
      <Card.Content>
      <Dropdown
        inline
        search
        options={currencies}
        defaultValue={currencies[0].value}
      />
        <Card.Meta content='Musicians' />
        <Card.Description content='Jake is a drummer living in New York.' />
      </Card.Content>
    </Card>
  )
}

export default ExchangeCard
