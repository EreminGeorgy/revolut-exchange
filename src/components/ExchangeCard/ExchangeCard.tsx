import { Card, Dropdown, Label } from 'semantic-ui-react'
import { DropdownProps } from 'semantic-ui-react/dist/commonjs/modules/Dropdown'
import { useContext, FC, SyntheticEvent, useCallback, useMemo } from 'react'

import { CURRENCIES, currenciesData } from '../../constants/currencies'
import { CurrencyDataContext } from '../../contexts/CurrencyDataContext'
import { SET_MAIN_CURRENCY, SET_DEPENDENT_CURRENCY } from '../../reducers/currencyDataReducer/actions'
import { InputState, WalletsState } from '../types'
import { formatCurrencyString } from '../../util/formatCurrencyString'
import { DICTIONARY } from '../../constants/dictionary'
import CurrencyInput from '../CurrencyInput'
import { getComputedCurrency } from '../../util/getComputedCurrency'

import './ExchangeCard.css'



type Props = {
  isMain: boolean
  currencyValues: InputState
  setCurrencyValues: (value: InputState) => void
  wallets: WalletsState
  errorMessage: boolean
}

const ExchangeCard: FC<Props> = ({ isMain, currencyValues, setCurrencyValues, wallets, errorMessage }) => {

  const { state, dispatch, rate } = useContext(CurrencyDataContext)

  const cardCurrency = useMemo(
    () => isMain ? state.mainCurrency : state.dependentCurrency,
    [isMain, state.mainCurrency, state.dependentCurrency]
  )

  const isSelling = useMemo(
    () => isMain !== state.isBuyMode,
    [isMain, state.isBuyMode]
  )

  const actionType = isMain ? SET_MAIN_CURRENCY : SET_DEPENDENT_CURRENCY
  const handleDropdownChange = useCallback(
    (e: SyntheticEvent<HTMLElement, Event>, data: DropdownProps ) =>  {
      setCurrencyValues({
        main: '',
        dependent: '',
      })
      dispatch({ type: actionType, payload: data.value as CURRENCIES })
    },
    [dispatch, setCurrencyValues, actionType]
  )

  const handleInputChange = useCallback(
    (value: string) => {
      if (isMain) {
        setCurrencyValues({
          main: value,
          dependent: getComputedCurrency(value, rate, true),
        })
      } else {
        setCurrencyValues({
          main: getComputedCurrency(value, rate, false),
          dependent: value,
        })
      }
    },
    [isMain, setCurrencyValues, rate]
  )

  const balanceString = useMemo(
    () => {
      const currencyValue = isMain ? wallets[state.mainCurrency] : wallets[state.dependentCurrency]
      return `
        ${DICTIONARY.balance}
        ${formatCurrencyString(currencyValue.toFixed(2), cardCurrency)}
      `
    }
    ,
    [cardCurrency, state.mainCurrency, state.dependentCurrency, wallets, isMain]
  )

  formatCurrencyString('1', cardCurrency)

  return (
    <Card>
      <Card.Content >
        <div className="card-controls">
          <div>
            <Dropdown
              inline
              search
              options={currenciesData}
              value={cardCurrency}
              onChange={handleDropdownChange}
              data-testid={`${isMain ? 'currency-selector-main' : 'currency-selector-secondary'}`}
            />
          </div>
          <div>
            <CurrencyInput
              value={isMain ? currencyValues.main : currencyValues.dependent}
              handleInputChange={handleInputChange}
              testId={`${isMain ? 'currency-input-main' : ''}`}
            />
          </div>
        </div>
        <div>
          <Card.Meta content={balanceString} data-testid={`${isMain ? 'main-wallet' : ''}`}/>
          {errorMessage && isSelling && 
            <Label basic color='red' data-testid={`${isMain ? 'error-label-main' : ''}`}>
              {DICTIONARY.exeedsBallance}
            </Label>
          }
        </div>
      </Card.Content>
    </Card>
  )
}

export default ExchangeCard
