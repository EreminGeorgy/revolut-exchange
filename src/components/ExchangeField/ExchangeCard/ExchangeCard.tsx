import { Card, Dropdown, Input } from 'semantic-ui-react'
import { DropdownProps } from 'semantic-ui-react/dist/commonjs/modules/Dropdown'
import { InputProps } from 'semantic-ui-react/dist/commonjs/elements/Input'
import { useContext, FC, SyntheticEvent } from 'react'

import { currenciesData } from '../../../constants/currencies'
import { CurrencyDataContext } from '../../../contexts/CurrencyDataContext'
import { SET_MAIN_CURRENCY, SET_DEPENDENT_CURRENCY } from '../../../reducers/currencyDataReducer/actions'
import { InputState } from '../types'
import NUMBER_INPUT_PATTERN from '../../../constants/numberInputPattern'
import './ExchangeCard.css'

type Props = {
  isMain: boolean
  currencyValues: InputState
  setCurrencyValues: (value: InputState) => void
}

const ExchangeCard: FC<Props> = ({ isMain, currencyValues, setCurrencyValues }) => {

  const { state, dispatch, rate } = useContext(CurrencyDataContext)

  const actionType = isMain ? SET_MAIN_CURRENCY : SET_DEPENDENT_CURRENCY
  const dropdownValue = isMain ? state.mainCurrency : state.dependentCurrency
  const handleDropdownChange = (e: SyntheticEvent<HTMLElement, Event>, data: DropdownProps ) =>  dispatch({ type: actionType, payload: data.value?.toString() || '' })

  const inputValue = isMain ? currencyValues.main : currencyValues.dependent
  const handleInputChange = (e: SyntheticEvent<HTMLElement, Event>, data: InputProps ) => {
    const match = data.value.match(NUMBER_INPUT_PATTERN);
    
    if (!match) {
      return
    }

    if (isMain) {
      setCurrencyValues({
        main: data.value,
        dependent: ((Number(data.value)*rate).toFixed(2)),
      })
    } else {
      setCurrencyValues({
        main: (Number(data.value) / rate).toFixed(2),
        dependent: data.value,
      })
    }
  }

  return (
    <Card>
      <Card.Content className="card-content">
        <div>
          <Dropdown
            inline
            search
            options={currenciesData}
            value={dropdownValue}
            onChange={handleDropdownChange}
          />
          <Card.Meta content='Musicians' />
        </div>
        <div>
          <Input
            className="currency-input"
            transparent 
            placeholder='0'  
            size='large'
            value={inputValue}
            onChange={handleInputChange}
          />
        </div>
      </Card.Content>
    </Card>
  )
}

export default ExchangeCard
