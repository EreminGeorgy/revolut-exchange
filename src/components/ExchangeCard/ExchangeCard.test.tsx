import { render } from '@testing-library/react'

import { CURRENCIES } from '../../constants/currencies';

import ExchangeCard from './ExchangeCard'

test('matches its snapshot', () => {
  const component = render(<ExchangeCard
    isMain={true}
    currencyValues={{ 
      main: '1',
      dependent: '0',
    }}
    setCurrencyValues={() => null}
    wallets = {{
      [CURRENCIES.EUR]: 1000,
      [CURRENCIES.USD]: 1000,
      [CURRENCIES.GBP]: 1000,
    }}
    errorMessage={true}
  />);
  expect(component).toMatchSnapshot();
});