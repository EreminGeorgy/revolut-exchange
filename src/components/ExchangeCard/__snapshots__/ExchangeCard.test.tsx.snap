// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches its snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui card"
      >
        <div
          class="content"
        >
          <div
            class="card-controls"
          >
            <div>
              <div
                aria-expanded="false"
                class="ui inline search dropdown"
                data-testid="currency-selector-main"
                role="combobox"
              >
                <input
                  aria-autocomplete="list"
                  autocomplete="off"
                  class="search"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  aria-atomic="true"
                  aria-live="polite"
                  class="divider text"
                  role="alert"
                >
                  <i
                    class="uk flag"
                  />
                  GBP
                </div>
                <i
                  aria-hidden="true"
                  class="dropdown icon"
                />
                <div
                  class="menu transition"
                  role="listbox"
                >
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="eu flag"
                    />
                    <span
                      class="text"
                    >
                      EUR
                    </span>
                  </div>
                  <div
                    aria-checked="true"
                    aria-selected="true"
                    class="active selected item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="uk flag"
                    />
                    <span
                      class="text"
                    >
                      GBP
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <i
                      class="us flag"
                    />
                    <span
                      class="text"
                    >
                      USD
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div>
                <div
                  class="ui large transparent input currency-input"
                  data-testid="currency-input-main"
                >
                  <input
                    placeholder="0"
                    type="text"
                    value="1"
                  />
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="meta"
              data-testid="main-wallet"
            >
              
      Balance: 
      £
      10.00
    
            </div>
            <div
              class="ui red basic label"
              data-testid="error-label-main"
            >
              exeeds ballance
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui card"
    >
      <div
        class="content"
      >
        <div
          class="card-controls"
        >
          <div>
            <div
              aria-expanded="false"
              class="ui inline search dropdown"
              data-testid="currency-selector-main"
              role="combobox"
            >
              <input
                aria-autocomplete="list"
                autocomplete="off"
                class="search"
                tabindex="0"
                type="text"
                value=""
              />
              <div
                aria-atomic="true"
                aria-live="polite"
                class="divider text"
                role="alert"
              >
                <i
                  class="uk flag"
                />
                GBP
              </div>
              <i
                aria-hidden="true"
                class="dropdown icon"
              />
              <div
                class="menu transition"
                role="listbox"
              >
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <i
                    class="eu flag"
                  />
                  <span
                    class="text"
                  >
                    EUR
                  </span>
                </div>
                <div
                  aria-checked="true"
                  aria-selected="true"
                  class="active selected item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <i
                    class="uk flag"
                  />
                  <span
                    class="text"
                  >
                    GBP
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <i
                    class="us flag"
                  />
                  <span
                    class="text"
                  >
                    USD
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div>
              <div
                class="ui large transparent input currency-input"
                data-testid="currency-input-main"
              >
                <input
                  placeholder="0"
                  type="text"
                  value="1"
                />
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="meta"
            data-testid="main-wallet"
          >
            
      Balance: 
      £
      10.00
    
          </div>
          <div
            class="ui red basic label"
            data-testid="error-label-main"
          >
            exeeds ballance
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
